File: packages/db-schema/src/schema/calendar-events.ts
import { sql } from 'drizzle-orm'
import { integer, sqliteTable, text, index } from 'drizzle-orm/sqlite-core'
import type { VectorClock, FieldClocks } from '@memry/contracts/sync-api'

export type CalendarEventResponseStatus =
  | 'needsAction'
  | 'declined'
  | 'tentative'
  | 'accepted'

export interface CalendarAttendee {
  email: string
  displayName?: string | null
  responseStatus?: CalendarEventResponseStatus | null
  optional?: boolean | null
  organizer?: boolean | null
  self?: boolean | null
}

export interface CalendarReminderOverride {
  method: 'email' | 'popup'
  minutes: number
}

export interface CalendarReminders {
  useDefault: boolean
  overrides: CalendarReminderOverride[]
}

export interface CalendarConferenceEntryPoint {
  entryPointType: string
  uri?: string | null
  label?: string | null
  pin?: string | null
  meetingCode?: string | null
  passcode?: string | null
  regionCode?: string | null
}

export interface CalendarConferenceData {
  conferenceId?: string | null
  conferenceSolution?: {
    key?: { type?: string | null } | null
    name?: string | null
    iconUri?: string | null
  } | null
  entryPoints?: CalendarConferenceEntryPoint[]
  notes?: string | null
}

export type CalendarVisibility = 'default' | 'public' | 'private' | 'confidential'

export const calendarEvents = sqliteTable(
  'calendar_events',
  {
    id: text('id').primaryKey(),
    title: text('title').notNull(),
    description: text('description'),
    location: text('location'),
    startAt: text('start_at').notNull(),
    endAt: text('end_at'),
    timezone: text('timezone').notNull().default('UTC'),
    isAllDay: integer('is_all_day', { mode: 'boolean' }).notNull().default(false),
    recurrenceRule: text('recurrence_rule', { mode: 'json' }).$type<Record<
      string,
      unknown
    > | null>(),
    recurrenceExceptions: text('recurrence_exceptions', { mode: 'json' }).$type<Array<
      Record<string, unknown>
    > | null>(),
    attendees: text('attendees', { mode: 'json' }).$type<CalendarAttendee[] | null>(),
    reminders: text('reminders', { mode: 'json' }).$type<CalendarReminders | null>(),
    visibility: text('visibility').$type<CalendarVisibility | null>(),
    colorId: text('color_id'),
    conferenceData: text('conference_data', { mode: 'json' }).$type<CalendarConferenceData | null>(),
    parentEventId: text('parent_event_id'),
    originalStartTime: text('original_start_time'),
    targetCalendarId: text('target_calendar_id'),
    archivedAt: text('archived_at'),
    clock: text('clock', { mode: 'json' }).$type<VectorClock>(),
    fieldClocks: text('field_clocks', { mode: 'json' }).$type<FieldClocks>(),
    syncedAt: text('synced_at'),
    createdAt: text('created_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`),
    modifiedAt: text('modified_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`)
  },
  (table) => [
    index('idx_calendar_events_start_at').on(table.startAt),
    index('idx_calendar_events_archived_at').on(table.archivedAt)
  ]
)

export type CalendarEvent = typeof calendarEvents.$inferSelect
export type NewCalendarEvent = typeof calendarEvents.$inferInsert

File: apps/desktop-tauri/src-tauri/src/db/calendar_events.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "camelCase")]
pub struct CalendarEvent {
    pub id: String,
    pub title: String,
    pub description: Option<String>,
    pub location: Option<String>,
    pub start_at: String,
    pub end_at: Option<String>,
    pub timezone: String,
    pub is_all_day: bool,
    pub recurrence_rule: Option<String>,
    pub recurrence_exceptions: Option<String>,
    pub archived_at: Option<String>,
    pub clock: Option<String>,
    pub synced_at: Option<String>,
    pub created_at: String,
    pub modified_at: String,
    pub target_calendar_id: Option<String>,
    pub field_clocks: Option<String>,
    pub attendees: Option<String>,
    pub reminders: Option<String>,
    pub visibility: Option<String>,
    pub color_id: Option<String>,
    pub conference_data: Option<String>,
    pub parent_event_id: Option<String>,
    pub original_start_time: Option<String>,
}

impl CalendarEvent {
    pub fn from_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Self> {
        Ok(Self {
            id: row.get("id")?,
            title: row.get("title")?,
            description: row.get("description")?,
            location: row.get("location")?,
            start_at: row.get("start_at")?,
            end_at: row.get("end_at")?,
            timezone: row.get("timezone")?,
            is_all_day: row.get::<_, i64>("is_all_day")? != 0,
            recurrence_rule: row.get("recurrence_rule")?,
            recurrence_exceptions: row.get("recurrence_exceptions")?,
            archived_at: row.get("archived_at")?,
            clock: row.get("clock")?,
            synced_at: row.get("synced_at")?,
            created_at: row.get("created_at")?,
            modified_at: row.get("modified_at")?,
            target_calendar_id: row.get("target_calendar_id")?,
            field_clocks: row.get("field_clocks")?,
            attendees: row.get("attendees")?,
            reminders: row.get("reminders")?,
            visibility: row.get("visibility")?,
            color_id: row.get("color_id")?,
            conference_data: row.get("conference_data")?,
            parent_event_id: row.get("parent_event_id")?,
            original_start_time: row.get("original_start_time")?,
        })
    }
}
